### ISSUE TITLE
Udacity Deep Learning Note

### ISSUE THREAD

**Issue by yuzhenning** [yuzhenning/python#3]
最简单的神经网络:
输出非0即1，输出单元的输入经过了一个激活函数 f(h) 在此处就是指阶跃函数。
输出单元返回的是 f(h) 的结果，其中 h 是输出单元的输入：
h=∑i*w_i*x_i
这个架构最酷的一点，也是使得神经网络可以实现的原因，就是激活函数 f(h) 可以是 任何函数，并不只是上面提到的阶跃函数。
常见激活函数还有对数几率（又称作 sigmoid），tanh 和 softmax。这节课中我们主要使用 sigmoid 函数：

sigmoid 函数值域是 0 到 1 之间，它的输出还可以被解释为成功的概率。实际上，用 sigmoid 函数作为激活函数的结果，跟对数几率回归是一样的。

简单网络练习:

接下来你要用 NumPy 来计算一个简单网络的输出，它有两个输入节点，一个输出节点，激活函数是 sigmoid。你需要做的有：

实现 sigmoid 激活函数
计算神经网络输出
sigmoid 函数公式是：

sigmoid(x)=1/(1+e^{-x})

指数你可以使用 NumPy 的指数函数 np.exp。

这个网络的输出为：

y=f(h)=sigmoid(∑_i​​ *w_​i *​​ x_​i​​ +b)

要计算加权求和，你可以让元素相乘再相加，或者使用 NumPy 的 点乘函数.

**Comment by yuzhenning** [comment]
“单层感知器”
output = sigmoid(np.dot(weights, inputs) +bias)
输出端为sigmoid函数，其中变量为：weights， inputs（数据裤）和bias。


import numpy as np

def sigmoid(x):
    res = 1/(1+np.exp(-x))
    return res
    # TODO: Implement sigmoid function
    # pass

inputs = np.array([0.7, -0.3])
weights = np.array([0.1, 0.8])
bias = -0.1

# TODO: Calculate the output
output = sigmoid(np.dot(weights, inputs) +bias)

print('Output:')
print(output)

**Comment by yuzhenning** [comment]
# 学习权重
你了解了如何使用感知器来构建 AND 和 XOR 运算，但它们的权重都是人为设定的。如果你要进行一个运算，例如预测大学录取结果，但你不知道正确的权重是什么，该怎么办？你要从样本中学习权重，然后用这些权重来做预测。

要了解我们将如何找到这些权重，可以从我们的目标开始考虑。我们想让网络做出的预测与真实值尽可能接近。为了能够衡量，我们需要有一个指标来了解预测有多差，也就是误差 (error)。一个普遍的指标是误差平方和 sum of the squared errors (SSE)：

E=1/2​​ ∑μ ​​ ∑​j [y_j^μ​​ −​y_j^​μ​​ ]^​2
​​ 
# 梯度下降

**Comment by yuzhenning** [comment]
# 梯度下降

如 Luis 所说，用梯度下降，我们通过多个小步骤来实现目标。在这个例子中，我们希望一步一步改变权重来减小误差。借用前面的比喻，误差就像是山，我们希望走到山下。下山最快的路应该是最陡峭的那个方向，因此我们也应该寻找能够使误差最小化的方向。我们可以通过计算误差平方的梯度来找到这个方向。

要计算变化率，我们要转向微积分，具体来说是导数。一个函数 f(x) 的导函数 f​′​(x) 给到你的是 f(x) 在 x 这一点的斜率。例如 x2 ，x2​​  的导数是 f′​​ (x)=2x。所以，在 x=2 这个点斜率 f′​​ (2)=4。画出图来就是：
梯度就是对多变量函数导数的泛化。我们可以用微积分来寻找误差函数中任意一点的梯度，它与输入权重有关，下一节你可以看到如何推导梯度下降的步骤。
下面我画了一个拥有两个输入的神经网络误差示例，相应的，它有两个权重。你可以将其看成一个地形图，同一条线代表相同的误差，较深的线对应较大的误差。

每一步，你计算误差和梯度，然后用它们来决定如何改变权重。重复这个过程直到你最终找到接近误差函数最小值的权重，即中间的黑点。


# 注意事项
因为权重会走向梯度带它去的位置，它们有可能停留在误差小，但不是最小的地方。这个点被称作局部最低点。如果权重初始值有错，梯度下降可能会使得权重陷入局部最优，例如下图所示。

有方法可以避免这一点，被称作 momentum.

**Comment by yuzhenning** [comment]
# 梯度下降： 代码
之前我们看到一个权重的更新可以这样计算：
Δw_i​ =ηδx_​i
这里 error term δ 是指δ =(y−​y​^​​ )f​′​​ (h)=(y−y​^​​ )f​′​​ (∑w_​i​​ x_​i​​)
记住，上面公式中 (y−y​^​​ ) 是输出误差，f′​​ (h) 是激活函数 f(h) 的导函数，我们把这个导函数称做输出的梯度。

现在假设只有一个输出单元，我来把这个写成代码。我们还是用 sigmoid 来作为激活函数 f(h)。

**Comment by yuzhenning** [comment]
# Defining the sigmoid function for activations 
# 定义 sigmoid 激活函数
def sigmoid(x):
    return 1/(1+np.exp(-x))

# Derivative of the sigmoid function
# 激活函数的导数
def sigmoid_prime(x):
    return sigmoid(x) * (1 - sigmoid(x))

# Input data
# 输入数据
x = np.array([0.1, 0.3])
# Target
# 目标
y = 0.2
# Input to output weights
# 输入到输出的权重
weights = np.array([-0.8, 0.5])

# The learning rate, eta in the weight step equation
# 权重更新的学习率
learnrate = 0.5

# the linear combination performed by the node (h in f(h) and f'(h))
# 输入和权重的线性组合
h = x[0]*weights[0] + x[1]*weights[1]
# or h = np.dot(x, weights)

# The neural network output (y-hat)
# 神经网络输出
nn_output = sigmoid(h)

# output error (y - y-hat)
# 输出误差
error = y - nn_output

# output gradient (f'(h))
# 输出梯度
output_grad = sigmoid_prime(h)

# error term (lowercase delta)
error_term = error * output_grad

# Gradient descent step 
# 梯度下降一步
del_w = [ learnrate * error_term * x[0],
          learnrate * error_term * x[1]]
# or del_w = learnrate * error_term * x

**Comment by yuzhenning** [comment]
# 练习代码：梯度下降

import numpy as np

def sigmoid(x):
    """
    Calculate sigmoid
    """
    return 1/(1+np.exp(-x))

def sigmoid_prime(x):
    """
    # Derivative of the sigmoid function
    """
    return sigmoid(x) * (1 - sigmoid(x))

learnrate = 0.5
x = np.array([1, 2, 3, 4])
y = np.array(0.5)

# Initial weights
weight = np.array([0.5, -0.5, 0.3, 0.1])

### Calculate one gradient descent step for each weight
### Note: Some steps have been consilated, so there are
###       fewer variable names than in the above sample code

# TODO: Calculate the node's linear combination of inputs and weights
h = np.dot(x , weight)

# TODO: Calculate output of neural network
nn_output = sigmoid(h)

# TODO: Calculate error of neural network
error = y-nn_output

# TODO: Calculate the error term
#       Remember, this requires the output gradient, which we haven't
#       specifically added a variable for.
output_grad = sigmoid_prime(h)
error_term = error * output_grad

# TODO: Calculate change in weights
del_w = learnrate * error_term * x

print('Neural Network output:')
print(nn_output)
print('Amount of Error:')
print(error)
print('Change in Weights:')
print(del_w)
